File: frontend/src/services/api/modelGateway.ts
import { apiClient } from './client'
import type {
  ModelConfig,
  ModelConfigCreatePayload,
  ModelConfigUpdatePayload,
  ModelTestResponse,
  Provider,
  ProviderCreatePayload,
  ProviderUpdatePayload,
} from '../../types/modelGateway'
import type { Capability } from '../../constants/modelCatalog'

export async function listProviders(): Promise<Provider[]> {
  const response = await apiClient.get('/providers')
  return response.data.data.providers as Provider[]
}

export async function createProvider(payload: ProviderCreatePayload): Promise<Provider> {
  const response = await apiClient.post('/providers', payload)
  return response.data.data.provider as Provider
}

export async function updateProvider(
  providerId: string,
  payload: ProviderUpdatePayload,
): Promise<Provider> {
  const response = await apiClient.patch(`/providers/${providerId}`, payload)
  return response.data.data.provider as Provider
}

export async function deleteProvider(providerId: string): Promise<void> {
  await apiClient.delete(`/providers/${providerId}`)
}

export async function listModels(capability?: Capability): Promise<ModelConfig[]> {
  const params = capability ? { capability } : undefined
  const response = await apiClient.get('/models', { params })
  return response.data.data.models as ModelConfig[]
}

export async function createModel(payload: ModelConfigCreatePayload): Promise<ModelConfig> {
  const response = await apiClient.post('/models', payload)
  return response.data.data.model as ModelConfig
}

export async function updateModel(
  modelId: string,
  payload: ModelConfigUpdatePayload,
): Promise<ModelConfig> {
  const response = await apiClient.patch(`/models/${modelId}`, payload)
  return response.data.data.model as ModelConfig
}

export async function deleteModel(modelId: string): Promise<void> {
  await apiClient.delete(`/models/${modelId}`)
}

export async function testModel(modelId: string): Promise<ModelTestResponse> {
  const response = await apiClient.post(`/models/${modelId}/test`)
  return response.data.data as ModelTestResponse
}

File: backend/app/api/routes/model_gateway.py
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.api.deps.auth import get_current_user
from app.core.db import get_db
from app.db.models.user import User
from app.schemas.model_gateway import (
    Capability,
    ModelConfigCreate,
    ModelConfigRead,
    ModelConfigUpdate,
    ProviderCreate,
    ProviderRead,
    ProviderUpdate,
)
from app.services.model_gateway_service import ModelGatewayService

router = APIRouter(tags=["model-gateway"])


# ---------- providers ----------


@router.get("/providers")
def list_providers(
    current_user: User = Depends(get_current_user), db: Session = Depends(get_db)
) -> dict:
    service = ModelGatewayService(db)
    items = service.list_providers(current_user.id)
    return {
        "success": True,
        "data": {
            "providers": [ProviderRead.model_validate(item).model_dump() for item in items]
        },
    }


@router.post("/providers")
def create_provider(
    payload: ProviderCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ModelGatewayService(db)
    provider = service.create_provider(current_user.id, payload)
    return {
        "success": True,
        "data": {"provider": ProviderRead.model_validate(provider).model_dump()},
    }


@router.patch("/providers/{provider_id}")
def update_provider(
    provider_id: str,
    payload: ProviderUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ModelGatewayService(db)
    provider = service.update_provider(current_user.id, provider_id, payload)
    return {
        "success": True,
        "data": {"provider": ProviderRead.model_validate(provider).model_dump()},
    }


@router.delete("/providers/{provider_id}")
def delete_provider(
    provider_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ModelGatewayService(db)
    service.delete_provider(current_user.id, provider_id)
    return {"success": True, "data": {"deleted": True}}


# ---------- models ----------


@router.get("/models")
def list_models(
    capability: Capability | None = Query(default=None),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ModelGatewayService(db)
    items = service.list_models(current_user.id, capability)
    return {
        "success": True,
        "data": {"models": [ModelConfigRead.model_validate(service.model_to_read_dict(m)).model_dump() for m in items]},
    }


@router.post("/models")
def create_model(
    payload: ModelConfigCreate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ModelGatewayService(db)
    model = service.create_model(current_user.id, payload)
    return {
        "success": True,
        "data": {"model": ModelConfigRead.model_validate(service.model_to_read_dict(model)).model_dump()},
    }


@router.patch("/models/{model_id}")
def update_model(
    model_id: str,
    payload: ModelConfigUpdate,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ModelGatewayService(db)
    model = service.update_model(current_user.id, model_id, payload)
    return {
        "success": True,
        "data": {"model": ModelConfigRead.model_validate(service.model_to_read_dict(model)).model_dump()},
    }


@router.delete("/models/{model_id}")
def delete_model(
    model_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ModelGatewayService(db)
    service.delete_model(current_user.id, model_id)
    return {"success": True, "data": {"deleted": True}}


@router.post("/models/{model_id}/test")
async def test_model(
    model_id: str,
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
) -> dict:
    service = ModelGatewayService(db)
    result = await service.test_model(current_user.id, model_id)
    return {"success": result["success"], "data": result}
